client/displayservers/Wayland/wayland.c

```diff
@@ -1031,28 +1198,90 @@ static const struct wl_data_offer_listener dataOfferListener = {
 static void dataDeviceHandleDataOffer(void * data,
     struct wl_data_device * dataDevice, struct wl_data_offer * offer)
 {
-  wcb.stashedType = LG_CLIPBOARD_DATA_NONE;
+  wcb.pendingType = LG_CLIPBOARD_DATA_NONE;
   wcb.isSelfCopy  = false;
   wl_data_offer_add_listener(offer, &dataOfferListener, NULL);
 }
 
-static void dataDeviceHandleSelection(void * data,
+static void clipboardReadCancel(struct ClipboardRead * data)
+{
+  waylandEpollUnregister(data->fd);
+  close(data->fd);
+  wl_data_offer_destroy(data->offer);
+  free(data->buf);
+  free(data);
+  wcb.currentRead = NULL;
+}
+
+static void clipboardReadCallback(uint32_t events, void * opaque)
+{
+  struct ClipboardRead * data = opaque;
+  if (events & EPOLLERR) 
+  {
+    clipboardReadCancel(data);
+    return;
+  }
+
+  ssize_t result = read(data->fd, data->buf + data->numRead, data->size - data->numRead);
+  if (result < 0)
+  {
+    DEBUG_ERROR("Failed to read from clipboard: %s", strerror(errno));
+    clipboardReadCancel(data);
+    return;
+  }
+
+  if (result == 0)
+  {
+    data->buf[data->numRead] = 0;
+    wcb.stashedType = data->type;
+    wcb.stashedSize = data->numRead;
+    wcb.stashedContents = data->buf;
+
+    waylandEpollUnregister(data->fd);
+    close(data->fd);
+    wl_data_offer_destroy(data->offer);
+    free(data);
```

Comment:
Duped code with `clipboardReadCancel`?

Reply:
This differs by the lack of `free(data->buf)`, but switched to `clipboardReadCancel` anyways, and added an extra flag to decide whether the buffer should be freed.